pkg/helm/charts.go

```diff
@@ -0,0 +1,289 @@
+	if err := h.updateCache(ctx, helmRepo); err != nil {
+		return nil, err
+	}
+
+	chart, err := h.loadChart(ctx, helmRepo, c)
+	if err != nil {
+		return nil, fmt.Errorf("loading %s from chart: %w", filename, err)
+	}
+
+	for _, v := range chart.Raw {
+		if v.Name == filename {
+			return v.Data, nil
+		}
+	}
+
+	return nil, fmt.Errorf("failed to find file: %s", filename)
+}
+
+func (h *RepoManager) updateCache(ctx context.Context, helmRepo *sourcev1beta1.HelmRepository) error {
+	entry, err := h.entryForRepository(ctx, helmRepo)
+	if err != nil {
+		return err
+	}
+
+	r, err := repo.NewChartRepository(entry, DefaultChartGetters)
+	if err != nil {
+		return err
+	}
+
+	r.CachePath = h.CacheDir
+	_, err = r.DownloadIndexFile()
+
+	return err
+}
+
+func (h *RepoManager) loadChart(ctx context.Context, helmRepo *sourcev1beta1.HelmRepository, c *ChartReference) (*chart.Chart, error) {
+	o, err := h.chartPathOptionsFromRepository(ctx, helmRepo, c)
+	if err != nil {
+		return nil, fmt.Errorf("failed to configure client: %w", err)
+	}
+
+	chartLocation, err := o.LocateChart(c.Chart, h.envSettings())
+	if err != nil {
+		return nil, fmt.Errorf("locating chart %q: %w", c.Chart, err)
+	}
+
+	chart, err := loader.Load(chartLocation)
+	if err != nil {
+		return nil, fmt.Errorf("failed to load chart %q: %w", c.Chart, err)
+	}
+
+	return chart, nil
+}
+
+func (h *RepoManager) chartPathOptionsFromRepository(ctx context.Context, helmRepo *sourcev1beta1.HelmRepository, c *ChartReference) (*action.ChartPathOptions, error) {
+	// TODO: This should probably use Verify: true
+	co := &action.ChartPathOptions{
+		RepoURL: helmRepo.Spec.URL,
+		Version: c.Version,
+	}
+
+	if helmRepo.Spec.SecretRef != nil {
+		username, password, err := credsForRepository(ctx, h.Client, helmRepo)
+		if err != nil {
+			return nil, err
+		}
+
+		co.Username = username
+		co.Password = password
+	}
+
+	return co, nil
+}
+
+func (h *RepoManager) entryForRepository(ctx context.Context, helmRepo *sourcev1beta1.HelmRepository) (*repo.Entry, error) {
+	entry := &repo.Entry{
+		Name: helmRepo.GetName() + "-" + helmRepo.GetNamespace(),
+		URL:  helmRepo.Spec.URL,
+	}
+
+	if helmRepo.Spec.SecretRef != nil {
+		username, password, err := credsForRepository(ctx, h.Client, helmRepo)
+		if err != nil {
+			return nil, err
+		}
+
+		entry.Username = username
+		entry.Password = password
+	}
+
+	return entry, nil
+}
+
+func credsForRepository(ctx context.Context, kc client.Client, hr *sourcev1beta1.HelmRepository) (string, string, error) {
+	var secret corev1.Secret
+	if err := kc.Get(ctx, types.NamespacedName{Name: hr.Spec.SecretRef.Name, Namespace: hr.Namespace}, &secret); err != nil {
+		return "", "", fmt.Errorf("repository authentication: %w", err)
+	}
+
+	return string(secret.Data["username"]), string(secret.Data["password"]), nil
+}
+
+func (h *RepoManager) envSettings() *cli.EnvSettings {
+	conf := cli.New()
```

Comment:
Not sure you want to do this as it has quite a lot of side effects and default assumptions that you likely not want to have in a service, even if they seem harmless at present: https://github.com/helm/helm/blob/main/pkg/cli/environment.go#L73

Reply:
@hiddeco can you expand? I'm not really familiar with what it does (Niki and I inherrited this code, so we don't understand all of it completely). what would you suggest as an alternative?

Reply:
The `cli` package in Helm contains configuration defaults with an assumption that it runs as a CLI on a machine. This means that e.g. cache path defaults try to create absolute paths for configuration fields relative to the user directory, and in addition, that a whole set of environment variables are taken into account as global configuration flags.

This means that unless you are very aware of what the action does with the configured defaults, you may end up with data magically stored somewhere, or things being unexpectedly taken into account.

The simplest alternative is to pass the `cli.EnvSettings` construct with no default values and limited config, the place where the returned object is used doesn't seem to use many of them: https://github.com/helm/helm/blob/main/pkg/action/install.go#L675

Reply:
Thanks! We've resolved it in https://github.com/weaveworks/weave-gitops/pull/1148/commits/c37413a9d2b91d05a6e2cb857fd848524998e966

lmk if that works :smile:

Reply:
If someone without context now tries to expand on the configuration for another Helm action, it should crash or obviously fail and require them to investigate instead of things seemingly working without them knowing about any hidden side effects, so this looks all good to me :100:

Think that the CI wants a `go fmt` to agree on the indentation of the fields however :grimacing: